quodlibet/quodlibet/ext/songsmenu/tapbpm.py

```diff
@@ -0,0 +1,193 @@
+            self.min = 0
+            self.max = 0
+            self.floating_bpm = 0.0
+            self.floating_square = 0.0
+        elif now > self.last:
+            # Use previous 5 values to average BPM
+            bpms = []
+            bpms.append(60.0 / (now - self.last))
+            # and four out of the list
+            for i in iter(range(1, 5)):
+                if len(self.last_times) <= i:
+                    break
+                bpms.append((i + 1) * 60.0 / (now - self.last_times[-i]))
+
+            self.bpm = sum(bpms) / len(bpms)
+
+            # Exponentially weighted floating average
+            weight = (1.0 / self.clicks) ** .5
+            if weight < self.min_weight:
+                weight = self.min_weight
+            self.floating_bpm = \
+                self.floating_bpm * (1.0 - weight) \
+                + self.bpm * weight
+            self.floating_square = \
+                self.floating_square * (1.0 - weight) \
+                + self.bpm * self.bpm * weight
+
+            if self.bpm < self.min or self.average_count == 0:
+                self.min = self.bpm
+            if self.bpm > self.max or self.average_count == 0:
+                self.max = self.bpm
+            self.bpms_sum += self.bpm
+            self.squares_sum += self.bpm * self.bpm
+            self.average_count += 1
+
+            # Update history
+            self.last_times = self.last_times[-(self.keep - 1):] + [self.last]
+            self.last_bpms = self.last_bpms[-(self.keep - 1):] + [self.bpm]
+            self.last_floating_bpms = \
+                self.last_floating_bpms[-(self.keep - 1):] \
+                + [self.floating_bpm]
+            self.last_floating_squares = \
+                self.last_floating_squares[-(self.keep - 1):] \
+                + [self.floating_square]
+
+        self.last = now
+        self.clicks += 1
+
+
+class TapBpm(SongsMenuPlugin):
+    PLUGIN_ID = "Tap BPM"
+    PLUGIN_NAME = _("Tap BPM")
+    PLUGIN_DESC = _("Tap BPM for the selected song.")
+    PLUGIN_ICON = Icons.EDIT
+    PLUGIN_VERSION = "0.1"
+
+    def plugin_song(self, song):
+        self._window = window = \
+            Dialog(title=_("Tap BPM"), parent=self.plugin_window)
+
+        window.add_button(_("_Cancel"), Gtk.ResponseType.CANCEL)
+        window.add_icon_button(_("_Save"), Icons.DOCUMENT_SAVE,
+                             Gtk.ResponseType.OK)
+
+        window.set_default_size(300, 100)
+        window.set_border_width(6)
+        window.connect('response', self.response)
+
+        self._panel = TapBpmPanel(window, song)
+        window.vbox.pack_start(self._panel, False, True, 0)
+
+        window.vbox.show_all()
+        window.present()
+
+    def response(self, win, response):
+        self._window.hide()
+        self._window.destroy()
+        del self._window
+
+        if response != Gtk.ResponseType.OK:
+            return
+
+        # Save metadata
+        self._panel.save()
```

Comment:
Not a bug, but try to save before destroying the window. After destroy() widgets shouldn't be used anymore and while it currently doesn't use it in save(), it might in the future.